File: bench/bench_ocaml/test/test_rational.ml
module Rational = Math_lib.Rational
module Test = Math_lib.Common.Test

let _test_rational ?(log = false) () =
  let a = Rational.new' (Z.of_int 6) in
  let a = Rational.inverse a in
  let value_resolved = Rational.resolve a (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int 166) in

  let a : Rational.t = { p = Z.of_int 1; q = Z.of_int 6 } in
  let value_resolved = Rational.resolve a (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int 166) in

  let a : Rational.t = { p = Z.of_int (-1); q = Z.of_int 6 } in
  let value_resolved = Rational.resolve a (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int (-166)) in

  let a : Rational.t = { p = Z.of_int 1; q = Z.of_int 3 } in
  let b : Rational.t = { p = Z.of_int 1; q = Z.of_int 2 } in
  let value : Rational.t = Rational.add a b in
  let () = assert (value = { p = Z.of_int 5; q = Z.of_int 6 }) in
  let value_resolved = Rational.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int 833) in

  let a : Rational.t = { p = Z.of_int 1; q = Z.of_int 3 } in
  let b : Rational.t = { p = Z.of_int 1; q = Z.of_int 2 } in
  let value : Rational.t = Rational.sub a b in
  let () = assert (value = { p = Z.of_int (-1); q = Z.of_int 6 }) in
  let value_resolved = Rational.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int (-166)) in

  let a : Rational.t = { p = Z.of_int 1; q = Z.of_int 3 } in
  let b : Rational.t = { p = Z.of_int 1; q = Z.of_int 2 } in
  let value : Rational.t = Rational.mul a b in
  let () = assert (value = { p = Z.of_int 1; q = Z.of_int 6 }) in
  let value_resolved = Rational.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int 166) in

  let a : Rational.t = { p = Z.of_int 1; q = Z.of_int 3 } in
  let b : Rational.t = { p = Z.of_int 1; q = Z.of_int 2 } in
  let value : Rational.t = Rational.div a b in
  let () = assert (value = { p = Z.of_int 2; q = Z.of_int 3 }) in
  let value_resolved = Rational.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int 666) in

  if log then Test.log "Test finished"

let all ?log () = _test_rational ?log ()

File: bench/bench_kotlin/src/test/kotlin/math/TestRational.kt
package math

import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.annotations.Scope
import org.openjdk.jmh.annotations.State
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals

@State(Scope.Benchmark)
open class TestRational {

    object Rational : math.Rational

    @Test
    fun testScientific00() {
        val a = Rational.inverse(Rational.new(6))
        val resolved = Rational.resolve(a)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(166))
    }

    @Test
    fun testScientific01() {
        val a = Rational.inverse(Rational.new(6))
        val b = Rational.mul(Rational.new(-1))(a)
        val resolved = Rational.resolve(b)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(-166))
    }

    @Test
    fun testScientific02() {
        val a = Rational.inverse(Rational.new(3))
        val b = Rational.inverse(Rational.new(2))
        val v = Rational.add(a)(b)
        val resolved = Rational.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(833))
    }

    @Test
    fun testScientific03() {
        val a = Rational.inverse(Rational.new(3))
        val b = Rational.inverse(Rational.new(2))
        val v = Rational.sub(a)(b)
        val resolved = Rational.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(-166))
    }

    @Test
    fun testScientific04() {
        val a = Rational.inverse(Rational.new(3))
        val b = Rational.inverse(Rational.new(2))
        val v = Rational.mul(a)(b)
        val resolved = Rational.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(166))
    }

    @Test
    fun testScientific05() {
        val a = Rational.inverse(Rational.new(3))
        val b = Rational.inverse(Rational.new(2))
        val v = Rational.div(a)(b)
        val resolved = Rational.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(666)) // Hell yeah
    }

    @Test
    fun testScientific06() {
        val a = Rational.inverse(Rational.new(2))
        val b = Rational.inverse(Rational.new(3))
        val v = Rational.modulo(a)(b)
        val resolved = Rational.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(166)) // Hell yeah
    }

    @Benchmark
    fun all() {
        testScientific00()
        testScientific01()
        testScientific02()
        testScientific03()
        testScientific04()
        testScientific05()
        testScientific06()
    }
}